dbshell: use sslmode and options from DATABASES['OPTIONS']

aviddiviner opened the pull request:
The latest [Cockroach docs for Django](https://www.cockroachlabs.com/docs/stable/connect-to-the-database.html?filters=python&filters=django) advise setting connection information like so:

```
DATABASES = {
    'default': {
        'ENGINE': 'django_cockroachdb',
        'NAME': '{database}',
        'USER': '{username}',
        'PASSWORD': '{password}',
        'HOST': '{host}',
        'PORT': '{port}',
        'OPTIONS': {
            'sslmode': 'verify-full',
            'options': '--cluster={routing-id}'
        },
    },
}
```

This currently doesn't work, as the `OPTIONS` are never passed to the connection string. This pull request addresses that and allows connections to CockroachDB Serverless, that requires a `--cluster` routing key.

rafiss commented on django_cockroachdb/client.py:
```diff
@@ -17,15 +18,22 @@ def settings_to_cmd_args_env(cls, settings_dict, parameters):
+            # Use the default certs dir if sslmode is explicitly set.
+            url_params["sslmode"] = sslmode
+            insecure = False
         else:
```
i think it's best to be explicit. can we make this `elif sslmode == "disable" || sslmode == ""`, and then the `else` branch should show an error that an invalid sslmode was chosen

aviddiviner replied:
Sure, I'll add some more conditional checks for `sslmode` 👍

aviddiviner replied:
It's worth noting though, that currently we don't raise any errors on bad params, so I'm not sure if we want to change that behaviour. 🤷

timgraham replied:
Correct. I expect the server raises an error for invalid options.

aviddiviner replied:
I've made the change. It will pass it through now at least in the options if it is `'disable'`. And raise on any bad value.

aviddiviner replied:
Okay, I actually made a little tweak. I don't like raising on the value explicitly, it doesn't seem to fit in with the rest of the code, but I am always passing the value on through to `cockroach` now, so it will error out there. And I'm handling `sslmode=disable` nicely.

rafiss commented on django_cockroachdb/client.py:
```diff
@@ -36,11 +44,15 @@ def settings_to_cmd_args_env(cls, settings_dict, parameters):
         environ = os.environ.copy()
         if password:
-            environ['COCKROACH_URL'] = f'postgresql://{user}:{password}@{host}:{port}/{db}'
+            suffix = '?' + urllib.parse.urlencode(url_params) if url_params else ''
```
as-is, the `options` params will only be passed through if a password was used

i wonder if we should change this so COCKROACH_URL is _always_ used

aviddiviner replied:
I'm doing so now. 👍

rafiss reviewed (changes requested):
thanks for your contribution!

would you mind updating the commit and PR to clarify that this is specifically for using Django to connect with the `cockroach sql` shell?

most of the code lgtm, but i'd like @timgraham to weigh in as well

aviddiviner commented on django_cockroachdb/client.py:
```diff
@@ -17,16 +18,21 @@ def settings_to_cmd_args_env(cls, settings_dict, parameters):
-            # Default to insecure if no ca exists.
-            insecure = True
+        # Pass `sslmode` through with the connection URL, if given.
+        if sslmode in ["allow", "prefer", "require", "verify-ca", "verify-full", "disable"]:
```
I could just change this to `if sslmode:` so I just pass whatever value we get. Let me know if you prefer that @rafiss.

aviddiviner replied:
You know what, I prefer that actually. Makes the code more future proof, because the value can be handled by cockroach.

aviddiviner replied:
Well, I suppose they could introduce other SSL modes that we want `insecure` to be `True` for. I don't know. Anyway, it is what it is for now, I'll leave it over to you. 🤷

timgraham replied:
At first glance, `if sslmode:` looks nice to me.

timgraham commented:
Thanks. I'd prefer if the change to always use `COCKROACH_URL` be a separate commit and/or PR that precedes the new feature so as to (as much as possible) not mix refactors and new features. If you want to me to do that, I don't mind. I'm going to try to write some tests (and will do manual testing also).

As the patch stands now, I'm guessing there's no need to add `db`, `user`, `host`, and `port` to `args` since `COCKROACH_URL` is always set? Can `sslrootcert` and `insecure` be part of `url_params` instead of `args` too?

Have you tested the patch to confirm that `--cluster={routing-id}` works? I've only tested local connections in the past.

aviddiviner commented:
Okay, I've tried one last time now, to keep my changes pretty minimal and pretty close to the current behaviour, while still fixing the original issue – allowing us to use this with CockroachDB Serverless – and addresses most of the comments. I think other changes can happen in subsequent PRs. At least this is an improvement without changing any current behaviour.

@timgraham @rafiss Can you take a last look please?

aviddiviner commented:
> Have you tested the patch to confirm that `--cluster={routing-id}` works? I've only tested local connections in the past.

Yes, 100%. That's the reason for this PR. I'm using this in production as of now.

rafiss commented on django_cockroachdb/client.py:
```diff
@@ -36,11 +42,14 @@ def settings_to_cmd_args_env(cls, settings_dict, parameters):
         environ = os.environ.copy()
         if password:
-            environ['COCKROACH_URL'] = f'postgresql://{user}:{password}@{host}:{port}/{db}'
+            query = '?' + urllib.parse.urlencode(url_params) if url_params else ''
```
it still seems non-ideal that `options` are only passed through if a `password` is given. should we change the condition to `if password or options`?

timgraham replied:
I'm going to clean this up in a separate commit.

aviddiviner replied:
Okay yeah, I agree. I put in `or url_params`. I also squashed all the commits on this branch, so it's only a single commit merge.

timgraham replied:
Is `{user}:` (with a blank password) a valid URL? Anyway, you don't need to spend more time on this unless you want to take up my plan to do the refactor of always passing COCKRAOCH_URL in one commit, then add the new functionality in a second. That's my plan. I started writing some tests yesterday.

aviddiviner replied:
I'd like you guys to merge this if possible. I don't really want to refactor, just want a package that actually works with CockroachDB Serverless.

timgraham replied:
No worries, it won't delay this. The next release of this package is coming after CockroachDB 22.2 is released.

timgraham replied:
To answer my concern a couple comments above, a connection URL like `username:` (with no password) works fine and so does a colon followed by an empty port, so although the URL looks a little funny, there's no need to omit those colons if password or port are empty.

David, could you please check and test https://github.com/cockroachdb/django-cockroachdb/pull/269. After reviews, I'll merge the first commit in that PR and then move the second commit to this PR for that merge.

Thank you for your contribution, and sorry for the back and forth.

p.s. I've written [some unit tests](https://gist.github.com/timgraham/f9d21e09eb1d2b22db4b0dd9e7120458) to make future work here easier, although I need to figure out how I can include them in this repo while also having the Django test suite pick them up.
